### PR TITLE
Feat/1844 deactivate email

### LINKED ISSUE #1844: As a TDP user, I want to be notified when my account will soon be deactivated
**Description:**
Per [AC-02(03)](https://github.com/raft-tech/TANF-app/blob/develop/docs/Security-Compliance/Security-Controls/ac-2-3/index.md), TDP must deactivate user accounts after 180 days of inactivity. We want to notify users that their account will be deactivated **10 days prior to**  AND **3 days prior to** AND **1 day prior to** this deactivation. 

This deactivation warning email notification should be triggered when the time difference between the system's current datetime and the user's `last_login` equals 170 days, 177 days, 179 days. 

**Acceptance Criteria:** 

- [ ] User receives a `deactivation warning` email when they haven't logged into TDP in 170 days
- [ ] User receives a `deactivation warning` email when they haven't logged into TDP in 177 days
- [ ] User receives a `deactivation warning` email when they haven't logged into TDP in 179 days
- [ ] User does not receive a `deactivation warning` email at any other time interval. 
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
- [ ] set up logic for deactivation warning email trigger
- [ ] Run Testing Checklist and confirm all tests pass

**Supporting Documentation:**
- [HTML Email Template](https://reitermb.github.io/TDP-HTML-Email-Demo-Environment/)
- [deactivation warning email content](https://hackmd.io/L0uwkVgxSPqfLBwOcuozPg#Email-Notification-Trigger--When-an-account-has-been-inactive-for-Number-days)
- `last_login` field
deactivate_inactive

**Open Questions:**
- How hard would it be to send a warning email at 170 _and_ 178 days?  _not complex: can be handled in this ticket_
- Can we add more user metadata for sys admins to facilitate deactivation? 
  - number of days since last login
  - access to deactivation warning emails or indicator that tracks how many and what types of emails have been sent to the user
_tasking in the ES ticket can help with this. this may need to be spun off into a new ticket. flexible if complex._ 

**Notes:**
- depending on complexity, we may need to split this into 2 tickets: one for setting up the logic for the email trigger and another for email trigger
- Release 2.1

### PR DESCRIPTION
**Pull request by n0remac**
## Summary of Changes
_Provide a brief summary of changes_
Pull request closes #1844 

To test this branch I made the last login attribute editable in the DAC. These videos show me changing the date to trigger the emails. The celery-beat schedule  was set to every minute. 

https://user-images.githubusercontent.com/7119690/198750553-8ec2eed3-8641-48a4-8519-a8be311d1982.mov

https://user-images.githubusercontent.com/7119690/198750656-2eb4c2f8-2289-4434-8268-36d13646c13c.mov

## How to Test

This one is harder to test. The database will need to be modified and the cron schedule in common.py should be more frequent. I have not done this yet, it should still be done. I am pretty confident in the tests.

## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] User receives a `deactivation warning` email when they haven't logged into TDP in 170 days
- [x] User receives a `deactivation warning` email when they haven't logged into TDP in 177 days
- [x] User receives a `deactivation warning` email when they haven't logged into TDP in 179 days
- [x] User does not receive a `deactivation warning` email at any other time interval. 
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [x] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/email/helpers/test/test_account_deactivation_warnings.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    mocker.patch(
+        'tdpservice.email.helpers.account_deactivation_warning.automated_email.delay',
+        return_value=None
+    )
+    send_deactivation_warning_email([user], 10)
+
+    assert tdpservice.email.email.automated_email.delay.called_once_with(
+        recipient_email=user.email,
+        subject='Account Deactivation Warning: 10 days remaining',
+    )
```
Comment (raft-twhitlock) [inline tdrs-backend/tdpservice/email/helpers/test/test_account_deactivation_warnings.py:20]: For this unit test would it be worth it to mock a date and assert that the `deactivation_date` logic works? I can be swayed either way for the reason of why you also wouldn't test third party libs

Comment (n0remac) [reply tdrs-backend/tdpservice/email/helpers/test/test_account_deactivation_warnings.py:20]: Test like that are in another file. Test covers the functionality of send_deactivation_warning_email.

**Review by andrew-jameson — APPROVED** [review]

**Comment by ADPennington** [comment]
@n0remac will add the DAC log entries for these emails. Because all pilot users joined TDP in August 2022, its okay if this ticket is delayed until after 2.2. cc: @stevenino

**Comment by andrew-jameson** [comment]
Moved back for LogEntry functionality akin to #1845 per Sprint Review 10/25

**Comment by ADPennington** [comment]
@n0remac @jtimpe i merged develop into this branch and im having trouble getting owasp to run 😞 1845 was most recently merged into develop

**Comment by n0remac** [comment]
@ADPennington The screen recordings have been updated to show the logs.

**Comment by ADPennington** [comment]
this PR is primed for approval pending a copy of a deactivated warning email which you may forward to me. @n0remac @andrew-jameson @stevenino

**Review by ADPennington — APPROVED** [review]
thanks for the emails @n0remac! 🚀  this is ready to merge @andrew-jameson

**Pull request merged**